### PR TITLE
SONARJS-826 FP for S1854 when in a "finally" clause

### PR DESCRIPTION
**Pull request by ivandalbosco**

### CODE REVIEW

File: javascript-checks/src/test/resources/checks/DeadStore.js [hunk 1]
```diff
@@ -138,3 +138,40 @@ function assignment_in_lhs() {
...
+  var a = foo();                  // OK, dead store, but there is a try statement in the function
+  try {
+    var y = foo();                // OK, dead store, but there is a try statement in the function
+    try {
+      var z = foo();              // OK, dead store, but there is a try statement in the function
+    } finally {
+    }
+  } finally {
+  }
+}
```
Comment (vilchik-elena) [inline javascript-checks/src/test/resources/checks/DeadStore.js]: @ivandalbosco I think it's worthy to add try-statement in global scope here if you had IT failing due to that

File: javascript-checks/src/main/java/org/sonar/javascript/checks/DeadStoreCheck.java [hunk 2]
```diff
@@ -87,6 +88,12 @@ private void checkCFG(ControlFlowGraph cfg, FunctionTree functionTree) {
     Usages usages = lva.getUsages();
 
     for (CfgBlock cfgBlock : cfg.blocks()) {
+      if (isTryBlock(cfgBlock)) {
+        return;
+      }
+    }
+
+    for (CfgBlock cfgBlock : cfg.blocks()) {
```
Comment (SonarTech) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/DeadStoreCheck.java]: MAJOR Combine this loop with the one that starts on line 90. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS3047)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* MAJOR 1 major

Watch the comments in this conversation to review them.

File: javascript-checks/src/test/resources/checks/DeadStore.js (hunk 1 above)
Comment (ivandalbosco) [inline javascript-checks/src/test/resources/checks/DeadStore.js]: @vilchik-elena Thanks! With the new, simpler implementation, this won't even be needed.

File: javascript-checks/src/main/java/org/sonar/javascript/checks/DeadStoreCheck.java [hunk 3]
```diff
@@ -49,26 +50,26 @@
 
   @Override
   public void visitFunctionDeclaration(FunctionDeclarationTree tree) {
-    checkFunction(tree);
     super.visitFunctionDeclaration(tree);
+    checkFunction(tree);
```
Comment (vilchik-elena) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/DeadStoreCheck.java]: @ivandalbosco No need to change order we don't rely on visitor for finding "try"

File: javascript-checks/src/main/java/org/sonar/javascript/checks/DeadStoreCheck.java [hunk 4]
```diff
@@ -87,6 +88,12 @@ private void checkCFG(ControlFlowGraph cfg, FunctionTree functionTree) {
     Usages usages = lva.getUsages();
 
     for (CfgBlock cfgBlock : cfg.blocks()) {
+      if (isTryBlock(cfgBlock)) {
+        return;
```
Comment (vilchik-elena) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/DeadStoreCheck.java]: @ivandalbosco Minor but cheap to do: more this for-loop in the beginning of this function (thus we don't create useless LVA)

**Pull request merged**